Add metadata OBU encoding support

MaxValenzuela opened the pull request:
# Description
This is a first effort to add metadata OBU encoding support.

I have previously contributed a similar API for libaom (commit 1) so users can feel familiar with the usage to attach a metadata stream to a to-be-encoded frame.

Please review this approach as I am not totally familiarized with the packetization process and the last changes in packetization_kernel were to get the metadata OBU in the correct position so it relates to the frame it corresponds inside the stream, but maybe you can provide feedback. Also, had to increase the value in packetization_reorder_entry_ctor from 16 to 160 to be able to write the metadata as 16 wasn't enough. This number is arbitrary.

Testing this commit with ffmpeg generates a stream with metadata OBUs in same order and format of what we get from libaom.
The metadata writing calls right now are in the interest of adding HDR10 (MDCV/CLL) /HDR10+ (ITU-T.35) metadata.

Let me know any suggestion to improve this contribution so it follows your expectation the best.

Also, any suggestion on appropriate tests to create or run for this are very welcome

# Issue

<!--
Mention if the PR fixes or address an issue in this section
https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue
Example
Fixes #999
If this is a bug fix that does not have an issue created for it, please create one with enough info to reproduce the issue
--->

# Author(s)
@MaxValenzuela

# Performance impact
<!--
Type an x in the box that is relevant to your PR. Make sure to mention in what way in the description
Example
- [x] memory
--->
- [ ] quality
- [ ] memory
- [ ] speed
- [ ] 8 bit
- [ ] 10 bit
- [x ] N/A

# Test set
- [ ] obj-1-fast can be found [here](https://media.xiph.org/video/derf/objective-1-fast.tar.gz)
- [ ] other
- [x] N/A

# Merge method
- [x] Allow the maintainer to squash and merge when PR is ready to create a 1-commit to the master branch. The maintainer will be able to fix typos / combine commit messages to create a more readable 1-commit message or use whatever is stated in the 'Description' section
- [ ] I will clean up my commits and the maintainer shall use 'rebase and merge' to the master branch

1480c1 commented on Source/API/EbSvtAv1Metadata.h:
```diff
@@ -0,0 +1,121 @@
+
+struct EbBufferHeaderType;
+
+typedef enum {
+    OBU_METADATA_TYPE_AOM_RESERVED_0 = 0,
+    OBU_METADATA_TYPE_HDR_CLL = 1,
+    OBU_METADATA_TYPE_HDR_MDCV = 2,
+    OBU_METADATA_TYPE_SCALABILITY = 3,
+    OBU_METADATA_TYPE_ITUT_T35 = 4,
+    OBU_METADATA_TYPE_TIMECODE = 5,
+} OBU_METADATA_TYPE;
```
careful with unprefixed or generic prefixed enums since they add to the global namespace and can't be blocked or undone

hassount commented:
@MaxValenzuela thank you for the contribution. Would you be able to address the build issues of the PR before we dig into reviewing the code ?
Also, can you please sign the appropriate CLA if you / your company has not done so already ? https://github.com/AOMediaCodec/SVT-AV1#contributor-agreement

MaxValenzuela commented:
@hassount Hi, I've been having some issues running the tests locally, but I'm getting the hang of it. I'll push a new version when the issues are addressed and I can test them out.
As for the CLA, I have previously contributed on behalf of my company to libaom and I see the link takes you to the same AOM one. Let me know if or how I should provide you with proof about the signature.

hassount commented:
> @hassount Hi, I've been having some issues running the tests locally, but I'm getting the hang of it. I'll push a new version when the issues are addressed and I can test them out.
> As for the CLA, I have previously contributed on behalf of my company to libaom and I see the link takes you to the same AOM one. Let me know if or how I should provide you with proof about the signature.

Thank you @MaxValenzuela no need to sign another CLA.

hassount commented on Source/API/EbSvtAv1Metadata.h:
```diff
@@ -0,0 +1,115 @@
+#ifndef EbSvtAv1Metadata_h
```
Can you please add a license header to this file and all new files, if most of the code is already being ported from libaom, then please feel free to port the libaom license header.

hassount replied:
ping on this

hassount commented on Source/Lib/Encoder/Codec/EbEntropyCoding.c:
```diff
@@ -4113,6 +4113,21 @@ static void add_trailing_bits(struct AomWriteBitBuffer *wb) {
     }
 }
+
+static uint32_t write_obu_metadata(SvtMetadataT *metadata, uint8_t *const dst) {
```
can you please add a comment for this function to explain its purpose and dependencies ?

hassount commented on Source/Lib/Encoder/Codec/EbEntropyCoding.c:
```diff
@@ -4229,6 +4244,41 @@ static uint32_t write_frame_header_obu(SequenceControlSet *     scs_ptr,
     return total_size;
 }
 
+EbErrorType write_metadata_av1(Bitstream *bitstream_ptr, SvtMetadataArrayT *metadata,
```
can you please add a comment for this function to explain its purpose and dependencies ?

hassount commented on Source/Lib/Encoder/Codec/EbPacketizationReorderQueue.c:
```diff
@@ -22,6 +22,6 @@ EbErrorType packetization_reorder_entry_ctor(PacketizationReorderEntry* entry_pt
     entry_ptr->picture_number = picture_number;
     //16 should enough for show existing frame
-    EB_NEW(entry_ptr->bitstream_ptr, bitstream_ctor, 16);
+    EB_NEW(entry_ptr->bitstream_ptr, bitstream_ctor, 160);
```
this might be a significant byte overhead especially for low resolution low bitrate content. how many extra bytes do we actually need and for what metadata? can we  increase the allocation only when needed i.e. test on the option being enabled to start with.

MaxValenzuela replied:
An idea is that whenever the temporal entry has metadata, we can check the metadata size to see if 16 is enough to hold it and realloc to add the metadata size if needed.
The case where we need something something bigger is for dynamic tone mapping metadata (p.e. HDR10+) where every frame has its own related metadata.  
This check would be needed everytime dynamic metadata is about to be encoded since we can't predict the exact size or which frames will include it, if we want to keep it to the minimum.

MaxValenzuela replied:
My idea for this is reflected on the last commit of the latest PR, let me know what you think.

hassount replied:
sorry I seem to have missed this. anyways you seems to have addressed this already using the realloc. @1480c1 did we have an issue a while back with realloc (I may be not remembering correctly).

1480c1 replied:
Yes, we had a cppcheck issue of not checking the return of realloc to see if it was null or not

hassount commented on Source/Lib/Encoder/Globals/EbEncHandle.c:
```diff
@@ -3598,6 +3598,9 @@ static void copy_input_buffer(
     dst->pic_type = src->pic_type;
 
+    // Copy the metadata array
+    dst->metadata = src->metadata;
```
if this is added to the API and the buffer is filled at the application level, then every field needs to be copied separately and not the pointer redirected, because the pointer maybe recycled for the upcoming pictures.

MaxValenzuela replied:
Added the copy_metadata_buffer() function to attend these.

hassount commented on Source/Lib/Encoder/Globals/EbEncHandle.c:
```diff
@@ -3644,6 +3647,11 @@ static void copy_output_recon_buffer(
     dst->pic_type = src->pic_type;
+
+    // Copy the metadata array
+    dst->metadata = src->metadata;
```
if this is added to the API and the buffer is filled at the application level, then every field needs to be copied separately and not the pointer redirected, because the pointer maybe recycled for the upcoming pictures.

hassount commented on Source/Lib/Encoder/Globals/EbMetadataHandle.c:
```diff
@@ -0,0 +1,132 @@
+#include <stdlib.h>
```
Thank you for not adding this to EbEncHandle.c :) 
can you please add a license header to this ?

hassount commented on Source/Lib/Encoder/Globals/EbMetadataHandle.c:
```diff
@@ -0,0 +1,132 @@
+
+EB_API void print_metadata_array(SvtMetadataArrayT *metadata) {
+    if(!metadata){
+        printf("Metadata Struct is NULL\n");
```
can we use macros from EbLog.h in order to unify printfs and be able to mute them when needed ?

MaxValenzuela replied:
I think that was left out as it was being used for debugging, we can either remove it or just add the SVT_LOG as you mention if it can be helpful for the user using the API

MaxValenzuela replied:
Decided to remove this one as it was not important for the current PR, it can be added on a next PR if it's found useful, but was mostly for debugging.

hassount commented on Source/Lib/Encoder/Globals/EbMetadataHandle.c:
```diff
@@ -0,0 +1,132 @@
+        }
+    }
+    return sz;
+}
```
you might need an empty line at the end otherwise the style check would not pass.

MaxValenzuela commented:
Thank you for the review @hassount ! 
There is still an MSVC compiling issue on the checks, should I force push so there is no non-compiling commits on history? I see your guidelines mention to avoid it when addressing comments, but there is still that check issue.

hassount commented:
> Thank you for the review @hassount !
> There is still an MSVC compiling issue on the checks, should I force push so there is no non-compiling commits on history? I see your guidelines mention to avoid it when addressing comments, but there is still that check issue.

Yes, if you have to rebase there's no way around it :/ we need to have all tests passing, I ll review again once they are.

MaxValenzuela commented:
Since I'm adding a new pointer to the _EbBufferHeaderType_ to contain the metadata arrays, I think any test that makes use of the _EbBufferHeaderType_ struct and doesn't get updated to have the metadata pointer set to NULL in its initialization will be having issues. 

This is the same case for FFmpeg tests, we would need to modify their libsvtav1.c class which calls the SVT encoder with the same changes applied to EbAppProcessCmd.c in the first commit of this PR. 

Let me know if this stage is OK for a review from your side or if I should go further into this (updating many test cases) or what would be a good way to proceed so if this gets accepted in the near future, we can update FFmpeg at the same time to avoid compatibility issues.

Thank you!

hassount commented on Source/Lib/Encoder/Codec/EbPacketizationProcess.c:
```diff
@@ -725,12 +738,42 @@ void *packetization_kernel(void *input_ptr) {
+            svt_metadata_array_free(&pcs_ptr->parent_pcs_ptr->input_ptr->metadata);
+        } else {
+            // Copy metadata pointer to the queue entry related to current frame number
+            uint64_t                        currentPictureNumber = pcs_ptr->picture_number;
```
currentPictureNumber should be snake_case

hassount commented on Source/Lib/Encoder/Codec/EbPacketizationProcess.c:
```diff
@@ -740,9 +783,26 @@ void *packetization_kernel(void *input_ptr) {
                     output_stream_ptr);
 
         if (pcs_ptr->parent_pcs_ptr->has_show_existing) {
+            uint64_t                        nextPictureNumber = pcs_ptr->picture_number + 1;
```
snake_case

hassount commented:
> Since I'm adding a new pointer to the _EbBufferHeaderType_ to contain the metadata arrays, I think any test that makes use of the _EbBufferHeaderType_ struct and doesn't get updated to have the metadata pointer set to NULL in its initialization will be having issues.
> 
> This is the same case for FFmpeg tests, we would need to modify their libsvtav1.c class which calls the SVT encoder with the same changes applied to EbAppProcessCmd.c in the first commit of this PR.
> 
> Let me know if this stage is OK for a review from your side or if I should go further into this (updating many test cases) or what would be a good way to proceed so if this gets accepted in the near future, we can update FFmpeg at the same time to avoid compatibility issues.
> 
> Thank you!

Yes please try to wrap this up so that we can merge it before we move the repo to gitlab on Wednesday

MaxValenzuela commented:
As expected, the only failing test is ffmpeg as their libsvtav1.c would need to be updated to set the metadata pointer to a valid value.
I will push a new version that forces the metadata pointer on the buffer coming from API users to NULL so it is compatible and the test passes.

hassount commented:
> ll push a new version that forces the metadata pointer on the buffer coming from API users to NULL so it is compatible and the

yes, infact we need to try to make sure it's backward compatible and that proper null checks are done before accepting this.

MaxValenzuela commented:
> yes, infact we need to try to make sure it's backward compatible and that proper null checks are done before accepting this.

null checks should be working but I'm not sure how we can handle backward compatibility with libsvtav1.c since we are adding a new metadata pointer to EbBufferHeaderType that would be uninitialized.
All that really needs to be done on their side is setting that new metadata pointer member to NULL as I did here on EbAppProcessCmd.c to prevent the pointer to be uninitialized. 

I don't know if you have any recommendation for this or how API changes are usually handled.

silverbacknet commented:
FFmpeg doesn't has a good external library versioning system, unfortunately. Going forward you can set an SVTAV1_BUILD similar to how X264_BUILD, and refuse to build or work around missing features, but that doesn't help with someone using an older ffmpeg with a newer library. That's not a solvable problem until a functioning API versioning is put in place. The version macros in place in EbSvtAv1.h don't actually call out the API, just the release number, and the API changes separately.

Maybe time to just add that to the header?

hassount commented:
> he version macros in place in EbSvtAv1.h don't actually call out the API, just the release number, and the API changes separately.

can we check on the library version basically to force the pointer to be Null in the library if <v0.8.8 and it would work otherwise, we just have to do some tracking on when this would go to ffmpeg (and gstreamer potentially) to enable it. basically using the library versioning as the API versioning.

MaxValenzuela commented:
> can we check on the library version basically to force the pointer to be Null in the library if <v0.8.8 and it would work otherwise, we just have to do some tracking on when this would go to ffmpeg (and gstreamer potentially) to enable it. basically using the library versioning as the API versioning.

Do you mean ffmpeg should check the SVTAV1 version it's using to define if it should use the metadata pointer? Got a little confused there...
I wouldn't understand the usage of SVT checking its own API version to make the decision unless I'm missing something

hassount commented:
> > can we check on the library version basically to force the pointer to be Null in the library if <v0.8.8 and it would work otherwise, we just have to do some tracking on when this would go to ffmpeg (and gstreamer potentially) to enable it. basically using the library versioning as the API versioning.
> 
> Do you mean ffmpeg should check the SVTAV1 version it's using to define if it should use the metadata pointer? Got a little confused there...
> I wouldn't understand the usage of SVT checking its own API version to make the decision unless I'm missing something

it would be basically shutting off the feature before a certain revision until both ffmpeg and svt-av1 are updated but thinking more about it, we might as well set it to Null until ffmpeg is updated.
